client/asset/eth/multirpc.go

```diff
@@ -0,0 +1,1298 @@
+	unconfirmedReceiptExpiration = time.Minute
+	tipCapSuggestionExpiration   = time.Hour
+	ipcHost                      = "IPC"
+	brickedFailCount             = 100
+	providerDelimiter            = " "
+)
+
+// nonceProviderStickiness is the minimum amount of time that must pass between
+// requests to DIFFERENT nonce providers. If we use a provider for a
+// nonce-sensitive (NS) operation, and later have another NS operation, we will
+// use the same provider if < nonceProviderStickiness has passed.
+var nonceProviderStickiness = time.Minute
+
+// The suggested tip cap is expected to be very-slowly changing. We'll only
+// update once per tipCapSuggestionExpiration.
+type cachedTipCap struct {
+	cap   *big.Int
+	stamp time.Time
+}
+
+type combinedRPCClient struct {
+	*ethclient.Client
+	rpc *rpc.Client
+}
+
+type provider struct {
+	host    string
+	ec      *combinedRPCClient
+	ws      bool
+	tipCapV atomic.Value // *cachedTipCap
+
+	// tip tracks the best known header.
+	tip struct {
+		sync.RWMutex
+		header      *types.Header
+		headerStamp time.Time
+		failStamp   time.Time
+		// failCount tracked consecutive errors. If failCount exceeds
+		// brickedFailCount, the client is considered irrecoverable. Successful
+		// header requests reset failCount.
+		failCount int
+	}
+}
+
+func (p *provider) setTip(header *types.Header) {
+	p.tip.Lock()
+	p.tip.header = header
+	p.tip.headerStamp = time.Now()
+	p.tip.failStamp = time.Time{}
+	p.tip.failCount = 0
+	p.tip.Unlock()
+}
+
+// cachedTip retrieves the last known best header.
+func (p *provider) cachedTip() *types.Header {
+	stale := time.Second * 10
+	if p.ws {
+		stale = time.Minute * 2
+	}
+
+	p.tip.RLock()
+	defer p.tip.RUnlock()
+	if time.Since(p.tip.failStamp) < failQuarantine || time.Since(p.tip.headerStamp) > stale {
+		return nil
+	}
+	return p.tip.header
+}
+
+// setFailed should be called after a failed request, the provider is considered
+// failed for failQuarantine.
+func (p *provider) setFailed() {
+	p.tip.Lock()
+	p.tip.failStamp = time.Now()
+	p.tip.failCount++
+	p.tip.Unlock()
+}
+
+// failed will be true if setFailed has been called in the last failQuarantine,
+// or if the failCount limit has been surpassed.
+func (p *provider) failed() bool {
+	p.tip.Lock()
+	defer p.tip.Unlock()
+	return time.Since(p.tip.failStamp) < failQuarantine || p.tip.failCount > brickedFailCount
+}
+
+// bestHeader get the best known header from the provider, cached if available,
+// otherwise a new RPC call is made.
+func (p *provider) bestHeader(ctx context.Context, log dex.Logger) (*types.Header, error) {
+	// Check if we have a cached header.
+	if tip := p.cachedTip(); tip != nil {
+		log.Tracef("using cached header from %q", p.host)
+		return tip, nil
+	}
+
+	log.Tracef("fetching fresh header from %q", p.host)
+	hdr, err := p.ec.HeaderByNumber(ctx, nil /* latest */)
+	if err != nil {
+		return nil, fmt.Errorf("HeaderByHash error: %w", err)
+	}
+	p.setTip(hdr)
+	return hdr, nil
+}
```

Comment:
Error message should say `HeaderByNumber`.

But it looks like we also need something like the following to satisfy the `ethFetcher` interface:

```go
func (p *provider) headerByHash(ctx context.Context, hash common.Hash, log dex.Logger) (*types.Header, error) {
	log.Tracef("fetching block header for %v from %q", hash, p.host)
	hdr, err := p.ec.HeaderByHash(ctx, hash)
	if err != nil {
		return nil, fmt.Errorf("HeaderByHash error: %w", err)
	}
	return hdr, nil
}
```